In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm

data= pd.read_csv('2.02. Binary predictors.csv')
data

SAT Admitted  Gender
0    1363       No    Male
1    1792      Yes  Female
2    1954      Yes  Female
3    1653       No    Male
4    1593       No    Male
..    ...      ...     ...
163  1722      Yes  Female
164  1750      Yes    Male
165  1555       No    Male
166  1524       No    Male
167  1461       No    Male

[168 rows x 3 columns]

Tenemos 168 datos de personas que admitidas a la universidad según sus puntajes de SAT y su género, revisaremos las posibilidades de admisión según las variables presentadas y el modelo de regresión logistica.


Al usar más variables podemos obtener mayor precisión, podremos ver a continuacion las diferencias, primero usaremos solo la variable género y despues usaremos género y SAT...

In [2]:
df=data.copy()
df['Admitted']=df['Admitted'].map({'Yes':1, 'No':0})
df['Gender']=df['Gender'].map({'Female':1, 'Male':0})
df

SAT  Admitted  Gender
0    1363         0       0
1    1792         1       1
2    1954         1       1
3    1653         0       0
4    1593         0       0
..    ...       ...     ...
163  1722         1       1
164  1750         1       0
165  1555         0       0
166  1524         0       0
167  1461         0       0

[168 rows x 3 columns]

In [3]:
y= df['Admitted']
X= df['Gender']

In [4]:
x= sm.add_constant(X)
regresion=sm.Logit(y,x).fit()


Optimization terminated successfully.
         Current function value: 0.572260
         Iterations 5


In [5]:
regresion.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               Admitted   No. Observations:                  168
Model:                          Logit   Df Residuals:                      166
Method:                           MLE   Df Model:                            1
Date:                Tue, 11 Jul 2023   Pseudo R-squ.:                  0.1659
Time:                        17:21:31   Log-Likelihood:                -96.140
converged:                       True   LL-Null:                       -115.26
Covariance Type:            nonrobust   LLR p-value:                 6.283e-10
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.6436      0.222     -2.901      0.004      -1.078      -0.209
Gender         2.0786      0.363      5.727      0.000       1.367       2.790
==============================================================================
"""

In [6]:
np.exp(2.0786)

7.993270498536442

La posibilidad de admisión de una mujer es 7.99 veces mayor que un hombre, en este caso solo usamos la variable Genero...

A continuación emplearemos el Genero y el SAT...

In [7]:
df1=data.copy()
df1['Admitted']=df1['Admitted'].map({'Yes':1, 'No':0})
df1['Gender']=df1['Gender'].map({'Female':1, 'Male':0})
df1

SAT  Admitted  Gender
0    1363         0       0
1    1792         1       1
2    1954         1       1
3    1653         0       0
4    1593         0       0
..    ...       ...     ...
163  1722         1       1
164  1750         1       0
165  1555         0       0
166  1524         0       0
167  1461         0       0

[168 rows x 3 columns]

In [9]:
dep=df1['Admitted']
indep=df1[['Gender','SAT']]

In [10]:
ind=sm.add_constant(indep)
modelo=sm.Logit(dep,ind).fit()

Optimization terminated successfully.
         Current function value: 0.120117
         Iterations 10


In [11]:
modelo.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               Admitted   No. Observations:                  168
Model:                          Logit   Df Residuals:                      165
Method:                           MLE   Df Model:                            2
Date:                Tue, 11 Jul 2023   Pseudo R-squ.:                  0.8249
Time:                        17:42:16   Log-Likelihood:                -20.180
converged:                       True   LL-Null:                       -115.26
Covariance Type:            nonrobust   LLR p-value:                 5.118e-42
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -68.3489     16.454     -4.154      0.000    -100.598     -36.100
Gender         1.9449      0.846      2.299      0.022       0.287       3.603
SAT            0.0406      0.010      4.129      0.000       0.021       0.060
==============================================================================

Possibly complete quasi-separation: A fraction 0.27 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

Esta tabla de resultados nos permite ver que ahora la Log verosimilitud o Log likelihood es mucho mayor que cuando solo usamos la variable de Género para predecir la admisión a la universidad. Anteriomente fue el valor arrojado fue de  -96.140 y ahora es de -20.180, esto significa que la variable SAT es significativa para predecir los resultados.

In [12]:
np.exp(1.9449)

6.992932526814459

Lo que vemos ahora es que la posibilidad de admisión de una mujer es 7 veces mayor que un hombre en caso de que tengan el mismo puntaje del SAT.

En este caso vemos que la universidad permite a las mujeres, mayores posibilidades de entrar a la universidad, sin embargo, podriamos encontrarno en nuestro trabajo más variables que nos indiquen más criterios de admisión en la universidad.



**Precisión del modelo**

In [15]:
np.set_printoptions(formatter={'float':lambda x:"{0:0.2f}".format(x)})
modelo.predict()

array([0.00, 1.00, 1.00, 0.23, 0.02, 0.99, 1.00, 1.00, 1.00, 0.01, 1.00,
       1.00, 0.76, 0.00, 0.60, 1.00, 0.11, 0.12, 0.51, 1.00, 1.00, 1.00,
       0.00, 0.01, 0.97, 1.00, 0.48, 0.99, 1.00, 0.99, 0.00, 0.83, 0.25,
       1.00, 1.00, 1.00, 0.31, 1.00, 0.23, 0.00, 0.02, 0.45, 1.00, 0.00,
       0.99, 0.00, 0.99, 0.00, 0.00, 0.01, 0.00, 1.00, 0.92, 0.02, 1.00,
       0.00, 0.37, 0.98, 0.12, 1.00, 0.00, 0.78, 1.00, 1.00, 0.98, 0.00,
       0.00, 0.00, 1.00, 0.00, 0.78, 0.12, 0.00, 0.99, 1.00, 1.00, 0.00,
       0.30, 1.00, 1.00, 0.00, 1.00, 1.00, 0.85, 1.00, 1.00, 0.00, 1.00,
       1.00, 0.89, 0.83, 0.00, 0.98, 0.97, 0.00, 1.00, 1.00, 0.03, 0.99,
       0.96, 1.00, 0.00, 1.00, 0.01, 0.01, 1.00, 1.00, 1.00, 0.00, 0.00,
       0.02, 0.33, 0.00, 1.00, 0.09, 0.00, 0.97, 0.00, 0.75, 1.00, 1.00,
       0.01, 0.01, 0.00, 1.00, 0.00, 0.99, 0.57, 0.54, 0.87, 0.83, 0.00,
       1.00, 0.00, 0.00, 0.00, 1.00, 0.04, 0.00, 0.01, 1.00, 0.99, 0.52,
       1.00, 1.00, 0.05, 0.00, 0.00, 0.00, 0.68, 1.

En este arreglo se muetran los valores predichos por el modelo, lo que podemos leer es que aquellos tienen valores por debajo de 0.5 serian tendientes a 0, mientras lo que estan por encima, a 1, y eso quiere decir que los valores tendientes a 0 son aquellos no admitidos y lo 1 si admitidos

In [16]:
np.array(df1['Admitted'])

array([0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0,
       1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0,
       0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0], dtype=int64)

En este arreglo tenemos los valores originales que se entran en este en el data set. Si el 80% de estos datos coinciden con los datos predichos entonces quiere decir que hay un 80% de precisión

In [17]:
modelo.pred_table()

array([[69.00, 5.00],
       [4.00, 90.00]])

In [18]:
matriz=pd.DataFrame(modelo.pred_table())
matriz.columns=['Prediccion 0', 'Prediccion 1']
matriz=matriz.rename(index={0:'Actual 0', 1:'Actual 1'})
matriz

Prediccion 0  Prediccion 1
Actual 0          69.0           5.0
Actual 1           4.0          90.0

Esta es llamada una matriz de confusión, de ella podemos leer lo siguiente:

De 69 predicciones donde el valor fue 0, el valor real era 0

De 90 predicciones donde el valor fue 1, el valor real era 1

De 4 predicciones donde el valor fue 0, el valor real era 1

De 5 predicciones donde el valor fue 1, el valor real era 0

Entonces: En 159 casos el modelo fue correcto y en 9 caso se confundio, recordemos que el total de datos es de 168, por lo tanto, 159/168= 94,6 es el procentaje de precisión del modelo.

Hagamos una prueba, con un dataset de 19 registros, este fue separado 90-10..

In [19]:
test=pd.read_csv('2.03. Test dataset.csv')
test

SAT Admitted  Gender
0   1323       No    Male
1   1725      Yes  Female
2   1762      Yes  Female
3   1777      Yes    Male
4   1665       No    Male
5   1556      Yes  Female
6   1731      Yes  Female
7   1809      Yes  Female
8   1930      Yes  Female
9   1708      Yes    Male
10  1970      Yes  Female
11  2039      Yes  Female
12  1681      Yes  Female
13  1444       No    Male
14  1726       No    Male
15  1876      Yes  Female
16  1485       No    Male
17  1500       No  Female
18  1900      Yes    Male

In [20]:
test['Admitted']=test['Admitted'].map({'Yes':1, 'No':0})
test['Gender']=test['Gender'].map({'Female':1, 'Male':0})
test


SAT  Admitted  Gender
0   1323         0       0
1   1725         1       1
2   1762         1       1
3   1777         1       0
4   1665         0       0
5   1556         1       1
6   1731         1       1
7   1809         1       1
8   1930         1       1
9   1708         1       0
10  1970         1       1
11  2039         1       1
12  1681         1       1
13  1444         0       0
14  1726         0       0
15  1876         1       1
16  1485         0       0
17  1500         0       1
18  1900         1       0

In [23]:
ind

const  Gender   SAT
0      1.0       0  1363
1      1.0       1  1792
2      1.0       1  1954
3      1.0       0  1653
4      1.0       0  1593
..     ...     ...   ...
163    1.0       1  1722
164    1.0       0  1750
165    1.0       0  1555
166    1.0       0  1524
167    1.0       0  1461

[168 rows x 3 columns]

In [28]:
test_1=test['Admitted']
prueba=test.drop(['Admitted'], axis=1)
prueba= sm.add_constant(prueba)
prueba=prueba[ind.columns.values] #Fue necesario añadir esta linea de codigo porque no coincidia el orden de las columnas con en que esta arriba y es 100% necesario que esten iguales
prueba

const  Gender   SAT
0     1.0       0  1323
1     1.0       1  1725
2     1.0       1  1762
3     1.0       0  1777
4     1.0       0  1665
5     1.0       1  1556
6     1.0       1  1731
7     1.0       1  1809
8     1.0       1  1930
9     1.0       0  1708
10    1.0       1  1970
11    1.0       1  2039
12    1.0       1  1681
13    1.0       0  1444
14    1.0       0  1726
15    1.0       1  1876
16    1.0       0  1485
17    1.0       1  1500
18    1.0       0  1900

In [30]:
def precision (entradas, valoresReales, modelos):

    valoresPredecidos= modelos.predict(entradas) 
    arreglo=np.array([0,0.5,1])
    cm=np.histogram2d(valoresReales, valoresPredecidos, bins=arreglo)[0]
    matriz=(cm[0,0]+cm[1,1])/cm.sum()

    return cm, matriz


In [31]:
matrizConfusion=precision(prueba, test_1, modelo)
matrizConfusion

(array([[5.00, 1.00],
        [1.00, 12.00]]),
 0.8947368421052632)

Al hacer la prueba con datos que el modelo no habia contemplado lo que tenemos es que:

De 5 datos donde la prediccion fue 0, los datos reales eran 0

De 12 datos donde la prediccion fue 1, los datos reales fueron 1

De 1 dato donde la prediccion fue 1, los datos reales fueron 0

De 1 datos donde la prediccion fue 0, los datos reales fueron 1

Entonces hubo 17 aciertos y 2 errores, por lo tanto, la precisión ahora es de 89%. Además esto significa que disminuyo respecto a los valores que obtuvimos anteriormente, pues esta prueba implicó un sobreajuste al modelo.

In [33]:
matriz=pd.DataFrame(matrizConfusion[0])
matriz.columns=['Prediccion 0', 'Prediccion 1']
matriz=matriz.rename(index={0:'Actual 0', 1:'Actual 1'})
matriz

Prediccion 0  Prediccion 1
Actual 0           5.0           1.0
Actual 1           1.0          12.0